include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+       * that will be used in the definition of the functions that it is
+       * desired to compute the sensitivities of.
+       * @param[in] n_dependent_variables Number of scalar functions to be
+       * defined that will have a sensitivity to the given independent
+       * variables.
+       */
+      ADHelperBase(const unsigned int n_independent_variables,
+                   const unsigned int n_dependent_variables);
+
+      /**
+       * Destructor
+       */
+      virtual ~ADHelperBase();
+
+      //@}
+
+      /**
+       * @name Interrogation of internal information
+       */
+      //@{
+
+      /**
+       * Returns the number of independent variables that this object expects to
+       * work with.
+       */
+      std::size_t
+      n_independent_variables() const;
+
+      /**
+       * Returns the number of dependent variables that this object expects to
+       * operate on.
+       */
+      std::size_t
+      n_dependent_variables() const;
+
+      /**
+       * Prints the status of all queriable data.
+       *
+       * @param[in] stream The output stream to which the values are to be
+       * written.
+       */
+      void
+      print(std::ostream &stream) const;
+
+      /**
+       * Prints the values currently assigned to the independent variables.
+       *
+       * @param[in] stream The output stream to which the values are to be
+       * written.
+       */
+      void
+      print_values(std::ostream &stream) const;
+
+      /**
+       * Prints the statistics regarding the usage of the tapes.
+       *
+       * @param[in] stream The output stream to which the values are to be
+       * written.
+       * @param[in] tape_index The index of the tape to get the statistics of.
+       *
+       * @note This function only produces meaningful output when @p ad_type
+       * is a taped auto-differentiable number.
+       */
+      void
+      print_tape_stats(std::ostream &           stream,
+                       const types::tape_index &tape_index) const;
+
+      //@}
+
+      /**
+       * @name Operations specific to tapeless mode
+       */
+      //@{
+
+      /**
+       * Pre-specify the number of @p independent_variables to be used in
+       * tapeless mode.
+       *
+       * Although this function is called internally in the ADHelperBase
+       * constructor, there may be occasions when ADOL-C adtl::adoubles are
+       * created before an instance of this class is created. This function
+       * therefore allows one to declare at the earliest possible instance how
+       * many directional derivatives will be considered in tapeless mode.
+       *
+       * @warning With @p ensure_persistent_setting set to <tt>true</tt>,
+       * calling this function leaves the set number of directional
+       * derivatives in a persistent state, so it will not be possible to
+       * further modify during course of the program's execution.
```

Comment:
modify WHAT?

Reply:
I've extended the documentation to clarify this point.